subsys/net/l2/openthread/openthread.c

```diff
@@ -429,248 +269,50 @@ int openthread_send(struct net_if *iface, struct net_pkt *pkt)
-}
-
-int openthread_stop(struct openthread_context *ot_context)
-{
-	otError error;
-
-	if (IS_ENABLED(CONFIG_OPENTHREAD_COPROCESSOR)) {
-		return 0;
-	}
-
-	openthread_api_mutex_lock(ot_context);
-
-	error = otThreadSetEnabled(ot_context->instance, false);
-	if (error == OT_ERROR_INVALID_STATE) {
-		NET_DBG("Openthread interface was not up [%d]", error);
-	}
-
-	openthread_api_mutex_unlock(ot_context);
-
-	return 0;
-}
-
-static int openthread_init(struct net_if *iface)
-{
-	struct openthread_context *ot_context = net_if_l2_data(iface);
-	struct k_work_queue_config q_cfg = {
-		.name = "openthread",
-		.no_yield = true,
+	struct openthread_context *ot_l2_context = net_if_l2_data(iface);
+	struct openthread_state_changed_callback ot_l2_state_changed_cb = {
+		.openthread_state_changed_cb = ot_l2_state_changed_handler,
+		.user_data = (void *)ot_l2_context,
 	};
-	otError err = OT_ERROR_NONE;
-
-	NET_DBG("openthread_init");
-
-	k_mutex_init(&ot_context->api_lock);
-	k_work_init(&ot_context->api_work, openthread_process);
 
 	ll_addr = net_if_get_link_addr(iface);
 
-	openthread_api_mutex_lock(ot_context);
-
-	otSysInit(0, NULL);
-
-	ot_context->instance = otInstanceInitSingle();
-	ot_context->iface = iface;
-
-	__ASSERT(ot_context->instance, "OT instance is NULL");
-
-	if (IS_ENABLED(CONFIG_OPENTHREAD_SHELL)) {
-		platformShellInit(ot_context->instance);
-	}
-
-	if (IS_ENABLED(CONFIG_OPENTHREAD_COPROCESSOR)) {
-		err = otPlatUartEnable();
-		if (err != OT_ERROR_NONE) {
-			NET_ERR("Failed to enable UART: [%d]", err);
-		}
-
-		otNcpHdlcInit(ot_context->instance, ncp_hdlc_send);
-	} else {
-		otIp6SetReceiveFilterEnabled(ot_context->instance, true);
-		otIp6SetReceiveCallback(ot_context->instance,
-					ot_receive_handler, ot_context);
-
-#if defined(CONFIG_OPENTHREAD_NAT64_TRANSLATOR)
-
-		otIp4Cidr nat64_cidr;
+	/* To keep backward compatibility use the additional state change callback list from the ot
+	 * l2 context and register the callback to the openthread module.
+	 */
+	sys_slist_init(&ot_l2_context->state_change_cbs);
+	openthread_state_change_callback_register(&ot_l2_state_changed_cb);
 
-		if (otIp4CidrFromString(CONFIG_OPENTHREAD_NAT64_CIDR, &nat64_cidr) ==
-		    OT_ERROR_NONE) {
-			if (otNat64SetIp4Cidr(openthread_get_default_instance(), &nat64_cidr) !=
-			    OT_ERROR_NONE) {
-				NET_ERR("Incorrect NAT64 CIDR");
-			}
-		} else {
-			NET_ERR("Failed to parse NAT64 CIDR");
-		}
-		otNat64SetReceiveIp4Callback(ot_context->instance, ot_receive_handler, ot_context);
+	openthread_init(ot_receive_handler, (void *)ot_l2_context);
 
-#endif /* CONFIG_OPENTHREAD_NAT64_TRANSLATOR */
+	ot_l2_context->iface = iface;
 
-		sys_slist_init(&ot_context->state_change_cbs);
-		err = otSetStateChangedCallback(ot_context->instance,
-						&ot_state_changed_handler,
-						ot_context);
-		if (err != OT_ERROR_NONE) {
-			NET_ERR("Could not set state changed callback: %d", err);
-		}
+	openthread_mutex_lock();
```

Comment:
Is this mutex lock/unlock needed here? We're not calling any OT APIs here.

Reply:
That's right, removed.